[pytorch][codegen] migrate gen_variable_type to new data model

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49735 [pytorch][codegen] migrate gen_variable_type to new data model**

This is the final wave of autograd codegen data model migration.

After this PR:
- autograd codegen no longer depends on Declarations.yaml;
- autograd codegen sources are fully type annotated and pass mypy-strict check;

To avoid potential merge conflicts with other pending PRs, some structural
changes are intentionally avoided, e.g. didn't move inner methods out, didn't
change all inner methods to avoid reading outer function's variables, and etc.

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Confirmed clean mypy-strict run:
```
mypy --config mypy-strict.ini
```

Differential Revision: [D25678879](https://our.internmc.facebook.com/intern/diff/D25678879)

ljk53 committed:
[pytorch][codegen] migrate gen_variable_type to new data model

This is the final wave of autograd codegen data model migration.

After this PR:
- autograd codegen no longer depends on Declarations.yaml;
- autograd codegen sources are fully type annotated and pass mypy-strict check;

To avoid potential merge conflicts with other pending PRs, some structural
changes are intentionally avoided, e.g. didn't move inner methods out, didn't
change all inner methods to avoid reading outer function's variables, and etc.

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Confirmed clean mypy-strict run:
```
mypy --config mypy-strict.ini
```

[ghstack-poisoned]

ljk53 commented:
@albanD I also have it rebased on top of your PR: https://github.com/ljk53/pytorch/commit/313f5cc50b4c2eaa50128bd08d2efe19c998effa

The merge conflicts seem to be manageable - so I think this could go before or after your PR, let me know your plan :)

ljk53 committed:
Update on "[pytorch][codegen] migrate gen_variable_type to new data model"


This is the final wave of autograd codegen data model migration.

After this PR:
- autograd codegen no longer depends on Declarations.yaml;
- autograd codegen sources are fully type annotated and pass mypy-strict check;

To avoid potential merge conflicts with other pending PRs, some structural
changes are intentionally avoided, e.g. didn't move inner methods out, didn't
change all inner methods to avoid reading outer function's variables, and etc.

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Confirmed clean mypy-strict run:
```
mypy --config mypy-strict.ini
```

Differential Revision: [D25678879](https://our.internmc.facebook.com/intern/diff/D25678879)

[ghstack-poisoned]

albanD commented on tools/autograd/gen_variable_type.py:
```diff
@@ -274,203 +275,221 @@
+        # TODO: `cpp_type` is only to keep it byte-for-byte compatible with the old codegen, should remove.
+        # NB: This is not a clone of cpp.argument() - TensorOptionsArguments / faithful / binds are
+        # not handled properly as they are irrelevant for this codegen.
+        cpp_type = cpp.argumenttype_type(a.type, mutable=a.is_write, binds=a.name).cpp_type()
```
nit: `argumenttype_type` is a bit surprising in terms of naming convention :)

albanD commented on tools/autograd/gen_variable_type.py:
```diff
@@ -505,75 +524,76 @@ def guard_for(arg: SavedAttribute) -> Optional[str]:
-            if arg['type'] == 'Tensor' or arg['type'] == 'TensorList':
-                body.append('throw_error_for_complex_autograd({}, "{}");'.format(name, base_name))
+            name = arg.name
+            # TODO: should be `arg.type.is_tensor_like()`?
```
Well I guess an output can only be differentiable if it is tensor like. So, I would expect this check to not do much?

albanD commented on tools/autograd/gen_variable_type.py:
```diff
@@ -694,47 +712,48 @@ def wrap_output(return_values, var):
+                if is_tensor_list_type(return_info.type):
                     if base_name in MULTI_OUTPUT_SAFE_FUNCTIONS:
-                        creation_meta = "CreationMeta::MULTI_OUTPUT_SAFE"
+                        creation_meta = 'CreationMeta::MULTI_OUTPUT_SAFE'
```
Out of curiosity, is there a difference between `"` and `'` in python? Or is this just for consistency?

albanD commented on tools/autograd/gen_variable_type.py:
```diff
@@ -694,47 +712,48 @@ def wrap_output(return_values, var):
                     enforce_same_ptrs_stmts += [ENFORCE_SAME_TENSORLIST_STORAGE.substitute(tensorlist_name=arg),
                                                 ENFORCE_SAME_TENSORLIST_IMPL.substitute(tensorlist_name=arg)]
-                elif simple_type == 'Tensor':
+                elif noref_cpp_type == 'Tensor':
```
We cannot check `is_tensor_like()` here?

albanD commented on tools/autograd/gen_variable_type.py:
```diff
@@ -922,62 +928,47 @@ def dispatch_strategy(declaration):
-    simple_types = ', '.join(arg['simple_type'] for arg in arguments)
-    return f'{name}({simple_types})'
+def is_tensor_type(t: Type) -> bool:
+    # TODO: Should handle optional here?
```
Given the replacements above, it should not. Otherwise we should add an argument to disable them for most use cases here.

albanD reviewed (commented):
That looks quite good!
Small comments below.

ezyang reviewed (approved):
Approving to unblock

bhosmer commented on tools/autograd/gen_autograd.py:
```diff
@@ -177,11 +104,11 @@ def is_operator_selected_for_training(decl):
+        gen_variable_type(out, native_functions_path, differentiability_infos, template_path, operator_selector)
 
-        from . import gen_trace_type
+        from .gen_trace_type import gen_trace_type
```
It's probably worth pulling the inline imports (here and line 106, at least) out to the top if you're back in here before landing.

ezyang replied:
fwiw, the codegen is still simple enough and free of BC concerns that we should be able to come up with some organization that doesn't have circular problems. But I agree, it doesn't matter for codegen, inline import or whatever you want to fix the problem is fine.

bhosmer replied:
> Curious - what's the problem with inline import? I remember you mentioned we discovered some performance issue with inline import, but why it matters for codegen?

Oh definitely not a perf concern here (or most places), just hygiene.

bhosmer commented on tools/autograd/gen_variable_type.py:
```diff
@@ -274,203 +275,221 @@
+@dataclass(frozen=True)
+class NativeFunctionWithDifferentiabilityInfo:
+    func: NativeFunction
+    info: Optional[DifferentiabilityInfo]
```
note to self: understand why this needs to be optional

bhosmer replied:
Ah sure - might be worth pasting your answer in here as a comment, if you do another edit to this file.

bhosmer commented on tools/autograd/gen_variable_type.py:
```diff
@@ -922,62 +928,47 @@ def dispatch_strategy(declaration):
+        info, is_exact_match = find_info(f)
+        result.append(NativeFunctionWithDifferentiabilityInfo(
+            func=f,
+            info=info,
```
I could definitely be missing something obvious, but when might a `None` creep in here (necessitating its optionality)?

bhosmer commented on tools/jit/gen_unboxing_wrappers.py:
```diff
@@ -279,6 +280,66 @@ def argument_order(decl):
+
+def load_aten_declarations(path):
+    with open(path, 'r') as f:
+        declarations = yaml.load(f, Loader=YamlLoader)
```
@albanD JFYI after this PR the mainline autograd codegen no longer uses this at all, and after Sebastian's c10-full stuff it should be gone completely (along with `gen_unboxing_wrappers.py`)

bhosmer reviewed (approved):
Hey, looks good! Couple of very minor things noted inline, but not worth holding up landing, only if you're back in for some other reason.

(Though I don't know enough to know if Alban's question on `DifferentiableOutput` implies an issue.)

ljk53 commented on tools/jit/gen_unboxing_wrappers.py:
```diff
@@ -279,6 +280,66 @@ def argument_order(decl):
+        return returns[0]['type']
+    else:
+        return_types = [r['type'] for r in returns]
+        return 'std::tuple<{}>'.format(','.join(return_types))
```
This entire file is going to be deprecated and the modifications are all simple copy-paste.
I moved old flavor code from gen_autograd to here so all autograd code can pass mypy-strict check.

albanD replied:
This was not related to mypy actually. I meant in the generated string so that the c++ codegened code is easier to read

ljk53 committed:
Update on "[pytorch][codegen] migrate gen_variable_type to new data model"


This is the final wave of autograd codegen data model migration.

After this PR:
- autograd codegen no longer depends on Declarations.yaml;
- autograd codegen sources are fully type annotated and pass mypy-strict check;

To avoid potential merge conflicts with other pending PRs, some structural
changes are intentionally avoided, e.g. didn't move inner methods out, didn't
change all inner methods to avoid reading outer function's variables, and etc.

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Confirmed clean mypy-strict run:
```
mypy --config mypy-strict.ini
```

Differential Revision: [D25678879](https://our.internmc.facebook.com/intern/diff/D25678879)

[ghstack-poisoned]

albanD commented on tools/codegen/api/autograd.py:
```diff
@@ -87,3 +87,36 @@ class DifferentiabilityInfo:
+    # TODO: only to keep it byte-for-byte compatible with the old codegen, should remove.
+    cpp_type: str
+
+# Represents a differentiable `Return`.
```
My question is just about the comment.
But in here, when there is a `mul_` in native_functions but only `mul` defined in derivatives, we re-use the formula and do things like replacing "result" by "self".
It is just to try and capture in the comment the exact state that is saved here: before or after the adaptation based on the exact variant we're working with.

albanD commented on tools/codegen/api/types.py:
```diff
@@ -31,14 +31,16 @@ class BaseCType:
-    def cpp_type(self) -> str:
+    def cpp_type(self, *, strip_ref: bool = False) -> str:
+        if strip_ref:
+            return self.elem.cpp_type(strip_ref=strip_ref)
```
I just find the structure of the function hard to follow. Maybe an `else` statement would make it clearer?
But this is just a style thing, I don't think this the code is wrong.
But something like this is more natural to me:
```python
no_ref_type = self.elem.cpp_type()
if strip_ref:
  return no_ref_type
else:
  return f'const {no_ref_type} &'
```

ljk53 commented on tools/autograd/gen_variable_type.py:
```diff
@@ -959,62 +970,47 @@ def dispatch_strategy(declaration):
-        declaration['derivative'] = info if info and info.args_with_derivatives else None
-
-        # Currently, the '.strides()' to 'strides_or_error' replacement does not support
-        # 'self' derivatives of an inplace function, so we must check for this case.
```
oops - I think I missed this check while solving merge conflicts... sorry!

I had a revision with your PR fully patched which contains this check: https://github.com/ljk53/pytorch/commit/313f5cc50b4c2eaa50128bd08d2efe19c998effa#diff-f6553cec68caeaea36f6c8b14ff76a6d39dfd774e0ea9ef2f76e8d81fd9af5dfR1096-R1105

@albanD I can send out a PR to add this check back, or you can help fix it if you are about to land your final codegen PR - please let me know :)

albanD replied:
I can add that back sure. Thanks!
